### PR TITLE
[ConsensusHandler] accumulate consensus stats for each epoch

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

Collect two stats for each epoch:
- Narwhal certificates created per validator.
- Sui transactions submitted to consensus per validator

Create write mod for updating consensus indices per consensus commit, instead of per transaction.

Another change is to always set the `consensus_message_processed` bit regardless of transaction type. I think there was no dependency on some message types not setting `consensus_message_processed`, but let me know otherwise.

## Test Plan 

CI

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Review by akichidis — COMMENTED** [review]
Changes look good to me 👍  

Regarding this

> Another change is to always set the consensus_message_processed bit regardless of transaction type. I think there was no dependency on some message types not setting consensus_message_processed, but let me know otherwise.

by reviewing the code I don't see any immediate issue with it (in fact much cleaner approach) but I would defer to the other reviewers since I am not that familiar with it.

File: crates/sui-core/src/consensus_handler.rs
Comment (mwtian) [reply crates/sui-core/src/consensus_handler.rs:254]: `AuthorityIdentifier` is a Narwhal type. And I want to only use Narwhal types from Sui when interfacing with Narwhal, if possible.

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (mwtian) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1571]: Maybe it is unnecessary, but I'm trying to support rolling back binary as well, because this change is not guarded by protocol version. The cost should be minimal.

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (mwtian) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1818]: Before this change, I think no caller of `process_consensus_transactions_for_tests()` sets meaningful value for `ExecutionIndicesWithHash` in the fake `VerifiedSequencedConsensusTransaction`. Adding a comment.

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (lxfind) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:1834]: hmm I made an attempt to merge them in https://github.com/MystenLabs/sui/pull/14126, would that not work?

Comment (mwtian) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1834]: I think that should be fine. I don't think any test is checking the consensus indices and stats written after a consensus commit.

Comment (mwtian) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1834]: I will leave that refactor to your PR if it's ok :)

**Review by lxfind — APPROVED** [review]
Nice cleanup overall.

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (lxfind) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:127]: This type may grow in code, may be good to put in a separate file

Comment (mwtian) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:127]: Do we want to create a file to hold basic types from sui core authority, or just put this type in its own file? I think it will at most add a few more entries, e.g. total gas submitted by a validator.

File: crates/sui-core/src/consensus_handler.rs
Comment (mystenmark) [inline crates/sui-core/src/consensus_handler.rs:114]: Is this even necessary, now that the consensus handler is atomic?

File: crates/sui-core/src/authority/authority_per_epoch_store.rs
Comment (mwtian) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:127]: We can refactor later when the struct grows.

File: crates/sui-core/src/consensus_handler.rs
Comment (mwtian) [reply crates/sui-core/src/consensus_handler.rs:114]: Missed the comment before merging. This if block could be unnecessary, arguably along with some of the consensus_message_processed, processed_set & processed_cache logic. This can be done in a separately cleanup, because it seems riskier.

**Pull request merged**
